File: cpp/substring_with_concatenation_of_all_words.cc
/*
 * Filename: substring_with_concatenation_of_all_words.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Wed 19 Aug 2015 09:40:47 PM EDT
 */

#include <iostream>
#include <map>
#include <string>
#include <vector>

using namespace std;

class Solution {
	public:
		vector<int> findSubstring(string s, vector<string>& words) {
			vector<int> res;

			int m = s.length();
			int n = words.size();
			if (m == 0 || n == 0)
			{
				return res;
			}

			int l = words[0].length();
			if (l == 0)
			{
				return res;
			}

			map<string, int> mapping;
			for (auto w: words)
			{
				mapping[w]++;
			}

			int i = 0;
			while (i + n*l <= m)
			{
				map<string, int> mapping2;

				int j = i;
				while (j < m)
				{
					string sub = s.substr(j, l);
					if (mapping.find(sub) == mapping.end())
					{
						break;
					}
					else
					{
						mapping2[sub]++;
						if (mapping2[sub] > mapping[sub])
						{
							break;
						}

						j += l;
					}
				}

				if (j-i == n*l)
				{
					res.push_back(i);
				}

				i++;
			}

			return res;
		}
};

File: java/SubstringwithConcatenationofAllWords.java
/*
 * Filename: SubstringwithConcatenationofAllWords.java
 * Author: Jun Xu
 * Email: junx@bu.edu
 * Created Time: Fri 13 Mar 2015 10:36:38 PM EDT
 */

public class Solution {
	public List<Integer> findSubstring(String S, String[] L) {
		HashMap<String, Integer> map = new HashMap<String, Integer>();
		for (int i = 0; i < L.length; i++) {
			if (map.containsKey(L[i])) {
				map.put(L[i], map.get(L[i]) + 1);
			} else {
				map.put(L[i], 1);
			}
		}

		int l = S.length();
		int m = L.length;
		int n = L[0].length();

		List<Integer> ret = new ArrayList<Integer>();

		for (int i = 0; i <= l - m*n; i++) {
			if (match(S.substring(i, i+m*n), map, m, n)) {
				ret.add(i);
			}
		}

		return ret;
	}

	private boolean match(String S, HashMap<String, Integer> map, int m, int n) {
		HashMap<String, Integer> map2 = new HashMap<String, Integer>(map);

		int wordCnt = S.length() / n;
		String substr;

		for (int i = 0; i < wordCnt; i++) {
			substr = S.substring(i*n, (i+1)*n);
			if (map2.containsKey(substr)) {
				if (map2.get(substr) > 1) {
					map2.put(substr, map2.get(substr) - 1);
				} else {
					map2.remove(substr);
				}
			} else {
				break;
			}
		}

		if (map2.isEmpty()) {
			return true;
		} else {
			return false;
		}
	}
}
